File: apps/client-macos/Sources/TimeTrack/Sync/TimeEntryUploader.swift
import Foundation

/// The outcome of one upload attempt. `success` = the server upserted (idempotent on UUIDv7).
/// `permanent` = a non-401 4xx the record can never satisfy → drop it (don't wedge the queue).
/// `transient` = network error / 5xx → retry later with backoff. `authFailed` = a 401 that
/// survived one refresh-retry → stop; the session is likely invalid.
enum UploadResult: Equatable {
    case success
    case permanent(Int)
    case transient
    case authFailed
}

protocol Uploading {
    func upload(_ payload: Data) async -> UploadResult
}

/// PRD §7.5 — POSTs a buffered record payload to `<baseURL>/<path>` (default `time-entries`;
/// idle events pass `idle-events`) with the session bearer token. The API upserts on the
/// client-minted UUIDv7, so a retried record is a no-op. On a 401 it forces a token refresh and
/// retries once (mirrors PolicyClient/ProjectClient); a surviving 401 → authFailed. Not a capture
/// path — no AckGate.
final class TimeEntryUploader: Uploading {
    private let baseURL: URL
    private let session: AuthSession
    private let path: String

    init(baseURL: URL, session: AuthSession, path: String = "time-entries") {
        self.baseURL = baseURL
        self.session = session
        self.path = path
    }

    func upload(_ payload: Data) async -> UploadResult {
        do {
            let token = try await session.accessToken()
            let status = try await post(payload, token: token)
            if status == 401 {
                let refreshed = try await session.forceRefresh()
                let retried = try await post(payload, token: refreshed)
                return Self.classify(status: retried)   // a second 401 → .authFailed
            }
            return Self.classify(status: status)
        } catch {
            return .transient   // network error, refresh failure, etc. → retry later
        }
    }

    /// Pure status → result mapping (unit-tested; the async orchestration is build-verified).
    static func classify(status: Int) -> UploadResult {
        switch status {
        // Any 2xx = the server accepted the record. The activity-samples/batch endpoint
        // returns 202 Accepted; narrowing success to 200/201 made the client treat every
        // accepted batch as a transient failure, wedging the buffer and re-sending forever.
        case 200...299: return .success
        case 401: return .authFailed
        case 408, 429: return .transient
        case 500...599: return .transient
        case 400...499: return .permanent(status)
        default: return .transient
        }
    }

    private func post(_ payload: Data, token: String) async throws -> Int {
        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = payload
        let (_, response) = try await URLSession.shared.data(for: request)
        return (response as? HTTPURLResponse)?.statusCode ?? 0
    }
}

File: apps/client-windows/src/NiftyTimer/Sync/TimeEntryUploader.cs
using System.Net.Http;
using System.Net;
using System.Net.Http.Headers;
using NiftyTimer.Auth;

namespace NiftyTimer.Sync;

/// <summary>
/// PRD §7.5 — POSTs a buffered record payload to <c>&lt;baseUrl&gt;/&lt;path&gt;</c> (default
/// <c>time-entries</c>; idle events pass <c>idle-events</c>) with the session bearer token. The
/// API upserts on the client-minted UUIDv7, so a retried record is a no-op. On a 401 it forces a
/// token refresh and retries once; a surviving 401 → <see cref="UploadResult.AuthFailed"/>.
///
/// Not a capture path — no <see cref="Policy.AckGate"/>.
/// </summary>
public sealed class TimeEntryUploader : IUploader
{
    private readonly HttpClient _http;
    private readonly Uri _baseUrl;
    private readonly AuthSession _session;
    private readonly string _path;

    public TimeEntryUploader(HttpClient http, Uri baseUrl, AuthSession session, string path = "time-entries")
    {
        _http = http;
        _baseUrl = baseUrl;
        _session = session;
        _path = path;
    }

    /// <summary>
    /// Pure status → result mapping (unit-tested; the async orchestration is build-verified).
    ///
    /// Any 2xx is success. The activity-samples and idle-events endpoints answer 201, and the
    /// batch endpoint has answered 202 — narrowing success to 200/201 made the Mac client treat
    /// every accepted batch as a transient failure, wedging the buffer and re-sending forever.
    /// Do not narrow this range.
    ///
    /// 409 lands in <see cref="UploadResult.Permanent"/>: it means another machine holds the
    /// server's one-running-entry index, which no retry of THIS record can change. Callers that
    /// care about the difference read <see cref="UploadResult.Permanent.Status"/> — see
    /// <see cref="LiveEntryPublisher"/>. Safe to drop the record because the partial unique index
    /// only covers OPEN entries and the durable buffer only ever holds closed ones.
    /// </summary>
    public static UploadResult Classify(int status) => status switch
    {
        >= 200 and <= 299 => new UploadResult.Success(),
        401 => new UploadResult.AuthFailed(),
        408 or 429 => new UploadResult.Transient(),
        >= 500 and <= 599 => new UploadResult.Transient(),
        >= 400 and <= 499 => new UploadResult.Permanent(status),
        _ => new UploadResult.Transient(),
    };

    public async Task<UploadResult> UploadAsync(byte[] payload, CancellationToken cancellationToken = default)
    {
        try
        {
            var token = await _session.AccessTokenAsync(cancellationToken).ConfigureAwait(false);
            var status = await PostAsync(payload, token, cancellationToken).ConfigureAwait(false);
            if (status == (int)HttpStatusCode.Unauthorized)
            {
                var refreshed = await _session.ForceRefreshAsync(cancellationToken).ConfigureAwait(false);
                var retried = await PostAsync(payload, refreshed, cancellationToken).ConfigureAwait(false);
                return Classify(retried); // a second 401 → AuthFailed
            }

            return Classify(status);
        }
        catch (Exception e) when (e is AuthException or NotAuthenticatedException or HttpRequestException
                                      or TaskCanceledException)
        {
            return new UploadResult.Transient(); // network error, refresh failure, etc. → retry later
        }
    }

    private async Task<int> PostAsync(byte[] payload, string token, CancellationToken cancellationToken)
    {
        using var content = new ByteArrayContent(payload);
        content.Headers.ContentType = new MediaTypeHeaderValue("application/json");

        using var request = new HttpRequestMessage(HttpMethod.Post, new Uri(_baseUrl, _path))
        {
            Content = content,
        };
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        using var response = await _http.SendAsync(request, cancellationToken).ConfigureAwait(false);
        return (int)response.StatusCode;
    }
}
